cli/2024-06-04-aws-irsa-support.md

```diff
@@ -0,0 +1,376 @@
+        Ports:          9443/TCP, 9090/TCP
+        Host Ports:     0/TCP, 0/TCP
+        :
+        Environment:
+          UCP_CONFIG:                   /etc/config/ucp-config.yaml
+          BASE_PATH:                    /apis/api.ucp.dev/v1alpha3
+          TLS_CERT_DIR:                 /var/tls/cert
+          PORT:                         9443
+        Mounts:
+          /etc/config from config-volume (rw)
+          /var/run/secrets/eks.amazonaws.com/serviceaccount from aws-iam-token (ro)
+          /var/run/secrets/kubernetes.io/serviceaccount from kube-api-access-d22ld (ro)
+          /var/tls/cert from cert (ro)
+```
+
+Note: Since AWS IRSA works by mounting the service-account token to pods, the decision to use IRSA should be taken at install time. Credentials can be registered later on.
+
+**post installation**
+
+Once the installation is complete, the user can use rad env update to store relevant information for deploying resources to accounts and region associated with specific radius environment.
+
+```
+rad env update qa --aws-account-id <aws-account-id> --aws-region <aws-region> 
+```
+
+At this point, we can use aws AssumeRole to manage resources using environment specific account information. rp (for recipes) and ucp will have code changes that make sure of aws assumerole to be able to use the right credential specific to the radius environment for managing the aws resources.  
+
+```
+  client := sts.NewFromConfig(cfg)
+  assumeRoleArn := retrieveRoleARNCredential() //has logic to get the roleARN stored as credential
+  // Assume the new role
+  assumeRoleProvider := stscreds.NewAssumeRoleProvider(client, assumeRoleArn)
+  // Create a new config with the new credentials
+  cfg.Credentials = assumeRoleProvider
+  cfg.Region = region
+  // Create a new S3 service client
+  s3client := s3.NewFromConfig(cfg)
+  // Call the ListBuckets function
+  output, err := s3client.ListBuckets(context.TODO(), &s3.ListBucketsInput{})
+  :
+
+```
+
+### API design
+
+```
+@doc("AWS credential kind")
+enum AWSCredentialKind {
+  @doc("The AWS Access Key credential")
+  AccessKey,
+
+  @doc("The AWS IRSA credential")
+  IRSA,
+}
+```
+
+```
+model AwsIRSACredentialProperties extends AwsCredentialProperties {
+  @doc("Access Key kind")
+  kind: AWSCredentialKind.IRSA;
+
+  @doc("RoleARN for AWS IRSA identity")
+  @secret
+  roleARN: string;
+
+  @doc("The storage properties")
+  storage: CredentialStorageProperties;
+}
+```
+
+Based on this new model we have to update API calls. 
+
+### CLI Design
+
+* `rad init --full` should be updated to ask the user to choose between AWS access keys and IRSA. If IRSA is chosen, the user should be able to input a AWS IAM Role ARN. As a result of rad init, 
+   1. the helm charts should be updated with service-account token mounted as secret.
+   2. roleARN should be stored as a AWS credential
```

Comment:
> roleARN should be stored as a AWS credential

a bit confused by this statement.. what do we mean by roleARN as AWS credential?

Reply:
roleARN would be stored as the radius credential for AWS. Radius would assume this role in order to communicate with AWS. This would work since the cluster is configured with OIDC provider ,pods would have service-account token mounted, and role has the right trust entries to trust the radius service-accounts.